app/src/main/java/com/futurice/freesound/feature/common/ui/adapter/RecyclerViewAdapter.java

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright 2016 Futurice GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.futurice.freesound.feature.common.ui.adapter;
+
+import com.futurice.freesound.common.utils.AndroidPreconditions;
+import com.futurice.freesound.common.utils.Preconditions;
+import com.futurice.freesound.feature.common.DisplayableItem;
+import com.futurice.freesound.feature.common.scheduling.SchedulerProvider;
+import com.futurice.freesound.viewmodel.viewholder.BaseBindingViewHolder;
+
+import android.support.annotation.NonNull;
+import android.support.v7.util.DiffUtil;
+import android.support.v7.widget.RecyclerView;
+import android.view.ViewGroup;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+import io.reactivex.Observable;
+import timber.log.Timber;
+
+import static com.futurice.freesound.common.utils.Preconditions.get;
+
+/**
+ * Implementation of {@link RecyclerView.Adapter} for {@link DisplayableItem}.
+ */
+public final class RecyclerViewAdapter extends RecyclerView.Adapter {
+
+    @NonNull
+    private final List<DisplayableItem> modelItems = new ArrayList<>();
+
+    @NonNull
+    private final ItemComparator comparator;
+
+    @NonNull
+    private final Map<Integer, ViewHolderFactory> factoryMap;
+
+    @NonNull
+    private final Map<Integer, ViewHolderBinder> binderMap;
+
+    @NonNull
+    private final SchedulerProvider schedulerProvider;
+
+    public RecyclerViewAdapter(@NonNull final ItemComparator comparator,
+                               @NonNull final Map<Integer, ViewHolderFactory> factoryMap,
+                               @NonNull final Map<Integer, ViewHolderBinder> binderMap,
+                               @NonNull final SchedulerProvider schedulerProvider) {
+        this.comparator = comparator;
+        this.factoryMap = factoryMap;
+        this.binderMap = binderMap;
+        this.schedulerProvider = schedulerProvider;
+    }
+
+    @Override
+    public RecyclerView.ViewHolder onCreateViewHolder(final ViewGroup parent, final int viewType) {
+        return factoryMap.get(viewType).createViewHolder(get(parent));
+    }
+
+    @Override
+    public void onBindViewHolder(final RecyclerView.ViewHolder holder, final int position) {
+        final DisplayableItem item = modelItems.get(position);
+        binderMap.get(item.type()).bind(get(holder), item);
+    }
+
+    @Override
+    public void onViewRecycled(final RecyclerView.ViewHolder holder) {
```

Comment:
If there's something like, just do what you suggested: make the changes in another PR. Otherwise we'll have PRs open forever trying to find the perfect solution.